types.go

```diff
@@ -0,0 +1,296 @@
+package autopilot
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/raft"
+)
+
+//go:generate mockery -all -inpkg -case snake -testonly
+
+// RaftState is the status of a single server in the Raft cluster.
+type RaftState string
+
+const (
+	RaftNone     RaftState = "none"
+	RaftLeader   RaftState = "leader"
+	RaftVoter    RaftState = "voter"
+	RaftNonVoter RaftState = "non-voter"
+	RaftStaging  RaftState = "staging"
+)
+
+func (s RaftState) IsPotentialVoter() bool {
+	switch s {
+	case RaftVoter, RaftStaging, RaftLeader:
+		return true
+	default:
+		return false
+	}
+}
+
+// NodeStatus represents the health of a server as know to the autopilot consumer.
+// This should not take into account Raft health and the server being on a new enough
+// term and index.
+type NodeStatus string
+
+const (
+	NodeUnknown NodeStatus = "unknown"
+	NodeAlive   NodeStatus = "alive"
+	NodeFailed  NodeStatus = "failed"
+	NodeLeft    NodeStatus = "left"
+)
+
+type NodeType string
+
+const (
+	NodeVoter NodeType = "voter"
+)
+
+// Config represents all the tunables of autopilot
+type Config struct {
+	// CleanupDeadServers controls whether to remove dead servers when a new
+	// server is added to the Raft peers.
+	CleanupDeadServers bool
```

Comment:
All of these configurations existed in autopilot already and it wasn't an area I was fiddling with too much.

One thing I wanted though was to keep autopilot in control and not have the application needing to make decisions. Failed server removal is actually two kinds of removal. The first is modification of the raft configuration when the server is gone (such as after a force-leave -prune in Consul). The second is where we request the application to remove the server that has failed (which in Consul's case cases it to force-leave the server which eventually during the next round of reconciliation will get removed for being a stale raft server).

The calculations are not super simple and the goal here is to have all the logic about what can/should be removed remain in autopilot.

So we can certainly group the settings and document how they are related but I don't think they should live in the application.

Reply:
Also the reason for multiple methods on the interfaces is to keep clean the separation of calculating what things should be done from the logic that actually applies them. Originally there was no `promoter.CalculateFailedServerRemovals` method but then I realized our enterprise promoter would need a way to prevent server removals. So now the process is that the core autopilot module gathers all the potential removals, passes that through the `CalculateFailedServerRemovals` to filter them (going to change the name of this method now that I am thinking about it) and finally autopilot will take the filtered set of removals and remove everything in the proper order maintaining certain properties and preventing removals that would destabilize the cluster or would leave fewer voters than the user desires.